Digital Credentials: implement IPC layer

marcoscaceres opened the pull request:
#### ed88665aadd671ab377d7d39f1ad189b9fd9f033
<pre>
Digital Credentials: implement IPC layer
<a href="https://bugs.webkit.org/show_bug.cgi?id=277384">https://bugs.webkit.org/show_bug.cgi?id=277384</a>
<a href="https://rdar.apple.com/problem/132853809">rdar://problem/132853809</a>

Reviewed by Chris Dumez.

Implements the IPC layer for the Digital Credentials feature.
This allows requesting digital credentials, and handling cancellations.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/PlatformMac.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SmartPointerExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp: Added.
(WebKit::DigitalCredentialsCoordinatorProxy::DigitalCredentialsCoordinatorProxy):
(WebKit::DigitalCredentialsCoordinatorProxy::~DigitalCredentialsCoordinatorProxy):
(WebKit::DigitalCredentialsCoordinatorProxy::sharedPreferencesForWebProcess const):
(WebKit::DigitalCredentialsCoordinatorProxy::protectedPage const):
(WebKit::DigitalCredentialsCoordinatorProxy::requestDigitalCredential):
(WebKit::DigitalCredentialsCoordinatorProxy::cancel):
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.h: Added.
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.messages.in: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didAttachToRunningProcess):
(WebKit::WebPageProxy::dispatchActivityStateChange):
(WebKit::WebPageProxy::didDestroyFrame):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::resetState):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp: Added.
(WebKit::DigitalCredentialsCoordinator::DigitalCredentialsCoordinator):
(WebKit::DigitalCredentialsCoordinator::protectedPage const):
(WebKit::DigitalCredentialsCoordinator::requestDigitalCredential):
(WebKit::DigitalCredentialsCoordinator::cancel):
* Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteXRBindingProxy.cpp:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteXRBindingProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):

Canonical link: <a href="https://commits.webkit.org/282338@main">https://commits.webkit.org/282338@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c319dd340784730a129c777b0ecfda031e5bf068

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62855 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42211 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15451 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66875 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13459 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49898 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13743 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50687 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9291 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65924 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39241 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54444 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31366 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35935 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12335 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57482 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12107 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68570 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6801 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11746 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58004 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6833 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54504 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58196 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13943 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5682 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38031 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39111 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40222 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38853 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/CMakeLists.txt:
```diff
@@ -235,6 +236,7 @@ set(WebKit_MESSAGES_IN_FILES
 
     Shared/Notifications/NotificationManagerMessageHandler
 
+    UIProcess/DigitalCredentials
```
Should likely be:
```
UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy
```

Likely why it's not building for glib ports.

marcoscaceres replied:
ah, makes sense.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -10275,6 +10282,7 @@ void WebPageProxy::resetState(ResetStateReason resetStateReason)
 
 #if ENABLE(WEB_AUTHN)
     protectedWebsiteDataStore()->authenticatorManager().cancelRequest(internals().webPageID, std::nullopt);
+    // FIXME: Implement equivelent Digital Credential Manager https://webkit.org/b/277850
```
typo: "equivelent"
Also, WebKit comments should be sentences and end with a period.

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.h:
```diff
@@ -0,0 +1,51 @@
+    void requestDigitalCredential(const WebCore::LocalFrame&, const WebCore::DigitalCredentialRequestOptions&, WebCore::DigitalCredentialRequestCompletionHandler&&);
+    void cancel(CompletionHandler<void()>&&) final;
+
+    WebPage& m_webPage;
```
We should really not be adding new C++ references like this in new code. Please consider using CheckedRef or WeakRef.

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.h:
```diff
@@ -0,0 +1,51 @@
+
+class WebPage;
+
+class DigitalCredentialsCoordinator : public WebCore::CredentialRequestCoordinatorClient {
```
Can the class be marked as `final`.

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.h:
```diff
@@ -0,0 +1,72 @@
+
+using DigitalRequestCompletionHandler = CompletionHandler<void(const WebCore::ExceptionData&)>;
+
+class DigitalCredentialsCoordinatorProxy : public IPC::MessageReceiver {
```
Can the class be marked as `final`.

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.h:
```diff
@@ -0,0 +1,72 @@
+    void requestDigitalCredential(WebCore::FrameIdentifier, FrameInfoData&&, WebCore::DigitalCredentialRequestOptions&&, DigitalRequestCompletionHandler&&);
+    void cancel(CompletionHandler<void()>&&);
+
+    WebPageProxy& m_webPageProxy;
```
We should use WeakRef or CheckedRef. This isn't safe.

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,73 @@
+
+void DigitalCredentialsCoordinatorProxy::requestDigitalCredential(FrameIdentifier frameId, FrameInfoData&& frameInfo, DigitalCredentialRequestOptions&& options, DigitalRequestCompletionHandler&& handler)
+{
+    // Handle the request for a digital credential.
```
Shouldn't this be a `FIXME` comment?

cdumez commented on LayoutTests/ipc/request-digital-credentials.html:
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE html>
+<!-- webkit-test-runner [ IPCTestingAPIEnabled=true ] -->
```
A little surprising to see this kind of test for other things than security testing?
Isn't there a way to leverage your new IPC code from JS without relying on window.IPC?

marcoscaceres replied:
Yes, hoping to get details on that. It's currently not documented.

marcoscaceres replied:
To be clear, right now not really (apart form very basic abort controller, maybe)... I'm trying to work my way to that by implementing the clients and a mock service.

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.h:
```diff
@@ -0,0 +1,52 @@
+
+class DigitalCredentialsCoordinator final : public WebCore::CredentialRequestCoordinatorClient {
+public:
+    explicit DigitalCredentialsCoordinator(WebPage&);
```
I don't see this object getting constructed anywhere, we should probably do it in this PR.

marcoscaceres replied:
Now in WebPage.cpp below...

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.h:
```diff
@@ -0,0 +1,52 @@
+    void requestDigitalCredential(const WebCore::LocalFrame&, const WebCore::DigitalCredentialRequestOptions&, WebCore::DigitalCredentialRequestCompletionHandler&&);
+    void cancel(CompletionHandler<void()>&&) final;
+
+    WeakPtr<WebKit::WebPage> m_webPage;
```
Once this object gets constructed, we should look at the ownership model and see if m_webPage can actually be null. My bet is that it can't and we should be using a WeakRef instead of a WeakPtr.

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,64 @@
+
+void DigitalCredentialsCoordinator::requestDigitalCredential(const LocalFrame& frame, const DigitalCredentialRequestOptions& options, DigitalCredentialRequestCompletionHandler&& handler)
+{
+    auto webFrame = WebFrame::fromCoreFrame(frame);
```
auto -> RefPtr

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,64 @@
+    if (!m_webPage)
+        return;
+
+    m_webPage->sendWithAsyncReply(Messages::DigitalCredentialsCoordinatorProxy::Cancel(), WTFMove(handler));
```
We need a RefPtr or Ref to the WePage *on the stack* before calling sendWithAsyncReply on it, as per smart pointer guidelines

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,64 @@
+    if (!webFrame || !m_webPage)
+        return;
+
+    m_webPage->sendWithAsyncReply(Messages::DigitalCredentialsCoordinatorProxy::RequestDigitalCredential(webFrame->frameID(), webFrame->info(), options), WTFMove(handler));
```
We need a RefPtr or Ref to the WePage on the stack before calling sendWithAsyncReply on it, as per smart pointer guidelines

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,80 @@
+
+const SharedPreferencesForWebProcess& DigitalCredentialsCoordinatorProxy::sharedPreferencesForWebProcess() const
+{
+    auto page = protectedPage();
```
We don't need this local variable, just use `protectedPage()->` below.

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,80 @@
+
+DigitalCredentialsCoordinatorProxy::~DigitalCredentialsCoordinatorProxy()
+{
+    auto page = protectedPage();
```
We don't need this local variable, just use protectedPage()-> below.

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,80 @@
+DigitalCredentialsCoordinatorProxy::DigitalCredentialsCoordinatorProxy(WebPageProxy& webPageProxy)
+    : m_webPageProxy(webPageProxy)
+{
+    m_webPageProxy->legacyMainFrameProcess().addMessageReceiver(Messages::DigitalCredentialsCoordinatorProxy::messageReceiverName(), m_webPageProxy->webPageIDInMainFrameProcess(), *this);
```
protectedLegacyMainFrameProcess()

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,80 @@
+DigitalCredentialsCoordinatorProxy::~DigitalCredentialsCoordinatorProxy()
+{
+    auto page = protectedPage();
+    page->legacyMainFrameProcess().removeMessageReceiver(Messages::DigitalCredentialsCoordinatorProxy::messageReceiverName(), m_webPageProxy->webPageIDInMainFrameProcess());
```
protectedLegacyMainFrameProcess()

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,80 @@
+
+void DigitalCredentialsCoordinatorProxy::cancel(CompletionHandler<void()>&& completionHandler)
+{
+    completionHandler();
```
We likely need a FIXME comment here.

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.h:
```diff
@@ -0,0 +1,76 @@
+
+private:
+    // IPC::MessageReceiver.
+    void didReceiveMessage(IPC::Connection&, IPC::Decoder&) override;
```
override -> final

cdumez replied:
Still not fixed.

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.h:
```diff
@@ -0,0 +1,76 @@
+
+    Ref<WebPageProxy> protectedPage() const;
+
+    WeakRef<WebPageProxy> m_webPageProxy;
```
I think we can call this m_page.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6031,6 +6035,7 @@ void WebPageProxy::didDestroyFrame(IPC::Connection& connection, FrameIdentifier
 {
 #if ENABLE(WEB_AUTHN)
     protectedWebsiteDataStore()->authenticatorManager().cancelRequest(webPageIDInMainFrameProcess(), frameID);
+    // FIXME: Implement equivalent Digital Credential Manager https://webkit.org/b/277850.
```
`(https://webkit.org/b/277850)`

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6223,6 +6228,7 @@ void WebPageProxy::didStartProvisionalLoadForFrameShared(Ref<WebProcessProxy>&&
 
 #if ENABLE(WEB_AUTHN)
     protectedWebsiteDataStore()->authenticatorManager().cancelRequest(internals().webPageID, frameID);
+    // FIXME: Implement equivalent Digital Credential Manager https://webkit.org/b/277850.
```
`(https://webkit.org/b/277850)`

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -10293,6 +10300,7 @@ void WebPageProxy::resetState(ResetStateReason resetStateReason)
 
 #if ENABLE(WEB_AUTHN)
     protectedWebsiteDataStore()->authenticatorManager().cancelRequest(internals().webPageID, std::nullopt);
+    // FIXME: Implement equivalent Digital Credential Manager https://webkit.org/b/277850.
```
`(https://webkit.org/b/277850)`

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,64 @@
+using namespace WebCore;
+
+DigitalCredentialsCoordinator::DigitalCredentialsCoordinator(WebPage& webPage)
+    : m_webPage(webPage)
```
I think we can use m_page and page.

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.h:
```diff
@@ -0,0 +1,76 @@
+#include <WebCore/CredentialRequestOptions.h>
+#include <WebCore/DigitalCredentialRequestOptions.h>
+#include <WebCore/FrameIdentifier.h>
+#include <wtf/CheckedRef.h>
```
Doesn't seem needed.

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,68 @@
+    if (!webFrame)
+        return;
+
+    if (protectedPage())
```
`if (RefPtr page = m_page.get())`

cdumez replied:
Also, you're failing to call the completion handler in the `else` case.

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,68 @@
+        return;
+
+    if (protectedPage())
+        protectedPage()->sendWithAsyncReply(Messages::DigitalCredentialsCoordinatorProxy::RequestDigitalCredential(webFrame->frameID(), webFrame->info(), options), WTFMove(handler));
```
`page->`

We want to avoid repeated calls to `protectedPage()` as it causes ref-counting churn.

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,68 @@
+{
+    RefPtr webFrame = WebFrame::fromCoreFrame(frame);
+    if (!webFrame)
+        return;
```
You're failing to call the completion handler before returning.

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,68 @@
+
+void DigitalCredentialsCoordinator::cancel(CompletionHandler<void()>&& handler)
+{
+    if (protectedPage())
```
`if (RefPtr page = m_page.get())`

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,68 @@
+void DigitalCredentialsCoordinator::cancel(CompletionHandler<void()>&& handler)
+{
+    if (protectedPage())
+        protectedPage()->sendWithAsyncReply(Messages::DigitalCredentialsCoordinatorProxy::Cancel(), WTFMove(handler));
```
`page->`

cdumez replied:
Also, you're failing to call the completion handler in the else case.

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.h:
```diff
@@ -0,0 +1,54 @@
+
+    RefPtr<WebPage> protectedPage() const;
+
+    WeakPtr<WebKit::WebPage> m_page;
```
`WebKit::` is unnecessary.

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,79 @@
+
+DigitalCredentialsCoordinatorProxy::~DigitalCredentialsCoordinatorProxy()
+{
+    protectedPage()->protectedLegacyMainFrameProcess()->removeMessageReceiver(Messages::DigitalCredentialsCoordinatorProxy::messageReceiverName(), protectedPage()->webPageIDInMainFrameProcess());
```
Sorry if I caused this with an earlier review comment (I think I might have!) but we should cache page in a local variable rather than calling `protectedPage()` twice. This causes ref-counting churn.

Should be something like:
```
Ref page = m_page.get();
page->protectedLegacyMainFrameProcess()->removeMessageReceiver(Messages::DigitalCredentialsCoordinatorProxy::messageReceiverName(), page->webPageIDInMainFrameProcess());
```

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,79 @@
+DigitalCredentialsCoordinatorProxy::DigitalCredentialsCoordinatorProxy(WebPageProxy& webPageProxy)
+    : m_page(webPageProxy)
+{
+    protectedPage()->protectedLegacyMainFrameProcess()->addMessageReceiver(Messages::DigitalCredentialsCoordinatorProxy::messageReceiverName(), protectedPage()->webPageIDInMainFrameProcess(), *this);
```
Sorry if I caused this with an earlier review comment (I think I might have!) but we should cache page in a local variable rather than calling `protectedPage()` twice. This causes ref-counting churn.

Should be something like:
```
Ref page = m_page.get();
page->protectedLegacyMainFrameProcess()->addMessageReceiver(Messages::DigitalCredentialsCoordinatorProxy::messageReceiverName(), page->webPageIDInMainFrameProcess(), *this);
```

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,79 @@
+namespace WebKit {
+using namespace WebCore;
+
+DigitalCredentialsCoordinatorProxy::DigitalCredentialsCoordinatorProxy(WebPageProxy& webPageProxy)
```
webPageProxy -> page

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,79 @@
+    // For now, we will simply call the handler with a dummy exception data.
+    WebCore::ExceptionData exceptionData;
+
+    handler(exceptionData);
```
I suggest:
```
handler({ ExceptionCode::NotSupportedError });
```

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,79 @@
+
+DigitalCredentialsCoordinatorProxy::~DigitalCredentialsCoordinatorProxy()
+{
+    RefPtr page = protectedPage();
```
`Ref page = m_page.get();` like above. `protectedPage()` function is only useful to avoid local variables.

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,79 @@
+{
+    // FIXME: Handle the request for a digital credential.
+    // For now, we will simply call the handler with a dummy exception data.
+    handler({ ExceptionCode::NotSupportedError });
```
Looks like this doesn't build, you'll likely need:
```
handler({ ExceptionCode::NotSupportedError, { } });
```

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.h:
```diff
@@ -0,0 +1,75 @@
+struct SharedPreferencesForWebProcess;
+struct FrameInfoData;
+
+using DigitalRequestCompletionHandler = CompletionHandler<void(const WebCore::ExceptionData&)>;
```
Still not taking a `std::optional<ExceptionData>`? Seems odd to require an exception given that the common case will succeed.

marcoscaceres replied:
yeah, sorry.... my idea was just to do the exception part first. This will take actual response data, but I haven't specified what that looks like yet. My idea was to support only exceptions for now, and then bring in the response data.

cdumez replied:
Are you planning to use a `std::variant<ReponseData, ExceptionData>` in a follow-up?

marcoscaceres replied:
Yes, that's what I was thinking. I'll add a FIXME and file a bug for it.

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,76 @@
+    RefPtr page = protectedPage();
+    if (!webFrame || !page) {
+        LOG_ERROR("Unable to get frame or page");
+        handler(ExceptionData { ExceptionCode::UnknownError });
```
I think we usually use `InvalidStateError` for such errors.

marcoscaceres replied:
Yeah, I guess InvalidStateError would be ok... this is very similar to "non-fully active", for where we use InvalidStateError in this API.

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,76 @@
+
+void DigitalCredentialsCoordinator::cancel(CompletionHandler<void()>&& handler)
+{
+    RefPtr page = protectedPage();
```
`RefPtr page = m_page.get();`

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,80 @@
+
+const SharedPreferencesForWebProcess& DigitalCredentialsCoordinatorProxy::sharedPreferencesForWebProcess() const
+{
+    Ref page = m_page.get();
```
not needed, only used one time.

cdumez commented on Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp:
```diff
@@ -0,0 +1,80 @@
+const SharedPreferencesForWebProcess& DigitalCredentialsCoordinatorProxy::sharedPreferencesForWebProcess() const
+{
+    Ref page = m_page.get();
+    return page->protectedLegacyMainFrameProcess()->sharedPreferencesForWebProcess();
```
`protectedPage()->`

cdumez commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -0,0 +1,76 @@
+void DigitalCredentialsCoordinator::requestDigitalCredential(const LocalFrame& frame, const DigitalCredentialRequestOptions& options, DigitalCredentialRequestCompletionHandler&& handler)
+{
+    RefPtr webFrame = WebFrame::fromCoreFrame(frame);
+    RefPtr page = protectedPage();
```
`RefPtr page = m_page.get();`

As mentioned earlier, `protectedPage()` is only useful to avoid local variables, not to initialize local variables.

cdumez reviewed (approved):
LGTM with nits.

mwyrzykowski commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteXRBindingProxy.cpp:
```diff
@@ -28,6 +28,7 @@
 
 #if ENABLE(GPU_PROCESS)
 
+#include "RemoteDeviceProxy.h"
```
thanks for the unified build fixes!

marcoscaceres committed:
Digital Credentials: implement IPC layer
https://bugs.webkit.org/show_bug.cgi?id=277384
rdar://problem/132853809

Reviewed by Chris Dumez.

Implements the IPC layer for the Digital Credentials feature.
This allows requesting digital credentials, and handling cancellations.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/PlatformMac.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SmartPointerExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp: Added.
(WebKit::DigitalCredentialsCoordinatorProxy::DigitalCredentialsCoordinatorProxy):
(WebKit::DigitalCredentialsCoordinatorProxy::~DigitalCredentialsCoordinatorProxy):
(WebKit::DigitalCredentialsCoordinatorProxy::sharedPreferencesForWebProcess const):
(WebKit::DigitalCredentialsCoordinatorProxy::protectedPage const):
(WebKit::DigitalCredentialsCoordinatorProxy::requestDigitalCredential):
(WebKit::DigitalCredentialsCoordinatorProxy::cancel):
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.h: Added.
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.messages.in: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didAttachToRunningProcess):
(WebKit::WebPageProxy::dispatchActivityStateChange):
(WebKit::WebPageProxy::didDestroyFrame):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::resetState):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp: Added.
(WebKit::DigitalCredentialsCoordinator::DigitalCredentialsCoordinator):
(WebKit::DigitalCredentialsCoordinator::protectedPage const):
(WebKit::DigitalCredentialsCoordinator::requestDigitalCredential):
(WebKit::DigitalCredentialsCoordinator::cancel):
* Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteXRBindingProxy.cpp:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteXRBindingProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):

Canonical link: https://commits.webkit.org/282338@main
